File: app/src/main/cpp/wn-steam-client/src/ws_connection.cpp
#include "wn_steam/ws_connection.h"

#include <android/log.h>

#include <exception>

#include <ixwebsocket/IXWebSocket.h>
#include <ixwebsocket/IXWebSocketMessage.h>
#include <ixwebsocket/IXWebSocketMessageType.h>
#include <ixwebsocket/IXSocketTLSOptions.h>

namespace wn_steam {

namespace {

constexpr const char* kLogTag = "WnSteamWS";

#define WN_LOGE(...) __android_log_print(ANDROID_LOG_ERROR, kLogTag, __VA_ARGS__)
#define WN_LOGI(...) __android_log_print(ANDROID_LOG_INFO,  kLogTag, __VA_ARGS__)

// Steam normal-close code; matches RFC 6455 §7.4.1 §1000.
constexpr uint16_t kNormalCloseCode = 1000;

TransportDisconnectReason map_close_reason(uint16_t close_code, bool tls_handshake_failed) {
    if (tls_handshake_failed)                       return TransportDisconnectReason::TlsHandshakeFailed;
    if (close_code >= 1000 && close_code < 1016)    return TransportDisconnectReason::RemoteClose;
    return TransportDisconnectReason::Unknown;
}

template <typename Cb, typename... Args>
void safe_invoke(Cb& cb, Args&&... args) {
    if (!cb) return;
    try {
        cb(std::forward<Args>(args)...);
    } catch (const std::exception& e) {
        WN_LOGE("user callback threw std::exception: %s", e.what());
    } catch (...) {
        WN_LOGE("user callback threw unknown exception");
    }
}

}  // namespace

WsConnection::WsConnection() : ws_(std::make_unique<ix::WebSocket>()) {
    // WS-level ping interval. This is NOT needed for keepalive — Steam CM
    // uses its own app-layer CMsgClientHeartBeat — but IXWebSocket reuses
    // this value as the transport poll() timeout: WebSocketTransport::poll()
    // sets `lastingTimeoutDelayInMs = _pingIntervalSecs` while the socket is
    // OPEN. A value of 0 therefore makes the transport worker thread call
    // poll() with a 0 ms timeout, so ppoll() returns instantly and the run
    // loop busy-spins a full CPU core (measured: 100% of one core, ~99 C).
    // Keep it positive so poll() actually blocks. 60 s = one tiny ping frame
    // per minute, which is negligible next to the ~9 s app heartbeat that
    // already wakes the radio far more often.
    ws_->setPingInterval(60);

    // Do NOT add a subprotocol — Steam rejects WebSocket upgrades that
    // include Sec-WebSocket-Protocol headers it did not request.
}

WsConnection::~WsConnection() {
    if (ws_) {
        user_initiated_close_.store(true);
        ws_->stop(kNormalCloseCode, "client closing");
    }
}

void WsConnection::set_ca_bundle_path(const std::string& path) {
    ca_bundle_path_ = path;
}

bool WsConnection::connect(const std::string& url) {
    TransportState expected = TransportState::Disconnected;
    if (!state_.compare_exchange_strong(expected, TransportState::Connecting)) {
        return false;
    }

    user_initiated_close_.store(false);
    WN_LOGI("connecting WSS: %s (caFile=%s)",
            url.c_str(),
            ca_bundle_path_.empty() ? "<unset>" : ca_bundle_path_.c_str());

    ws_->setUrl(url);
    ws_->disableAutomaticReconnection();  // CMClient owns the reconnect policy.

    // TLS configuration. IXWebSocket consumes a single PEM via `caFile`.
    // If the caller has populated ca_bundle_path_, point libssl at it;
    // otherwise IXWebSocket defaults to "SYSTEM" which on Android will
    // fail. We still issue the connect attempt so the failure is observable
    // via the disconnect callback rather than a silent state stuck in
    // Connecting.
    ix::SocketTLSOptions tls_opts;
    tls_opts.tls = true;
    if (!ca_bundle_path_.empty()) {
        tls_opts.caFile = ca_bundle_path_;
    }
    ws_->setTLSOptions(tls_opts);

    ws_->setOnMessageCallback([this](const ix::WebSocketMessagePtr& msg) {
        if (!msg) return;
        switch (msg->type) {
            case ix::WebSocketMessageType::Open: {
                state_.store(TransportState::Connected);
                ConnectedCallback cb;
                {
                    std::lock_guard<std::mutex> lk(cb_mu_);
                    cb = on_connected_;
                }
                safe_invoke(cb);
                break;
            }
            case ix::WebSocketMessageType::Message: {
                if (msg->binary) {
                    MessageCallback cb;
                    {
                        std::lock_guard<std::mutex> lk(cb_mu_);
                        cb = on_message_;
                    }
                    if (cb) {
                        const auto* p = reinterpret_cast<const uint8_t*>(msg->str.data());
                        safe_invoke(cb, std::span<const uint8_t>(p, msg->str.size()));
                    }
                } else {
                    WN_LOGE("dropped non-binary WS message (text frames are not part of the Steam CM protocol)");
                }
                break;
            }
            case ix::WebSocketMessageType::Close: {
                state_.store(TransportState::Disconnected);
                DisconnectedCallback cb;
                {
                    std::lock_guard<std::mutex> lk(cb_mu_);
                    cb = on_disconnected_;
                }
                const auto reason = user_initiated_close_.load()
                    ? TransportDisconnectReason::UserInitiated
                    : map_close_reason(msg->closeInfo.code, /*tls_handshake_failed=*/false);
                safe_invoke(cb, reason, msg->closeInfo.reason);
                break;
            }
            case ix::WebSocketMessageType::Error: {
                state_.store(TransportState::Disconnected);
                const bool tls_fail =
                    msg->errorInfo.reason.find("tls") != std::string::npos ||
                    msg->errorInfo.reason.find("TLS") != std::string::npos ||
                    msg->errorInfo.reason.find("SSL") != std::string::npos ||
                    msg->errorInfo.reason.find("certificate") != std::string::npos;
                DisconnectedCallback cb;
                {
                    std::lock_guard<std::mutex> lk(cb_mu_);
                    cb = on_disconnected_;
                }
                const auto reason = user_initiated_close_.load()
                    ? TransportDisconnectReason::UserInitiated
                    : map_close_reason(0, tls_fail);
                safe_invoke(cb, reason, msg->errorInfo.reason);
                break;
            }
            default:
                break;  // Ping/Pong/Fragment — not visible to upper layers.
        }
    });

    ws_->start();
    return true;
}

bool WsConnection::send(std::span<const uint8_t> data) {
    if (state_.load() != TransportState::Connected) return false;
    if (!ws_) return false;
    // IXWebSocket expects the payload as std::string; reinterpret-cast is
    // safe because std::string is contiguous and we treat it as a byte
    // container. IXWebSocket itself returns success=false if the socket
    // closed mid-call (handles the race against the worker thread's state
    // transitions), so no extra synchronization is needed here.
    const ix::WebSocketSendInfo info =
        ws_->sendBinary(std::string(reinterpret_cast<const char*>(data.data()),
                                    data.size()));
    return info.success;
}

void WsConnection::disconnect() {
    TransportState s = state_.load();
    if (s == TransportState::Disconnected) return;
    state_.store(TransportState::Disconnecting);
    user_initiated_close_.store(true);
    if (ws_) ws_->stop(kNormalCloseCode, "client disconnect");
    state_.store(TransportState::Disconnected);
}

TransportState WsConnection::state() const {
    return state_.load();
}

void WsConnection::set_on_message(MessageCallback cb) {
    std::lock_guard<std::mutex> lk(cb_mu_);
    on_message_ = std::move(cb);
}

void WsConnection::set_on_connected(ConnectedCallback cb) {
    std::lock_guard<std::mutex> lk(cb_mu_);
    on_connected_ = std::move(cb);
}

void WsConnection::set_on_disconnected(DisconnectedCallback cb) {
    std::lock_guard<std::mutex> lk(cb_mu_);
    on_disconnected_ = std::move(cb);
}

}  // namespace wn_steam

File: app/src/main/cpp/wn-steam-client/rust/src/ws_connection.rs
use crate::transport::{Transport, TransportDisconnectReason, TransportState};
use std::sync::atomic::{AtomicBool, AtomicU8, Ordering};
use std::sync::{mpsc, Arc, Mutex};
use std::thread::{self, JoinHandle};
use std::time::Duration;
use tungstenite::error::Error as WsError;
use tungstenite::stream::MaybeTlsStream;
use tungstenite::{connect, Message};

const NORMAL_CLOSE_CODE: u16 = 1000;
const READ_POLL_TIMEOUT: Duration = Duration::from_millis(100);

type MessageCallback = Arc<dyn Fn(&[u8]) + Send + Sync + 'static>;
type ConnectedCallback = Arc<dyn Fn() + Send + Sync + 'static>;
type DisconnectedCallback = Arc<dyn Fn(TransportDisconnectReason, &str) + Send + Sync + 'static>;

pub struct WsConnection {
    shared: Arc<WsConnectionShared>,
    sender: Mutex<Option<mpsc::Sender<WsCommand>>>,
    worker: Mutex<Option<JoinHandle<()>>>,
}

#[derive(Default)]
struct WsConnectionShared {
    state: AtomicU8,
    user_initiated_close: AtomicBool,
    ca_bundle_path: Mutex<String>,
    on_message: Mutex<Option<MessageCallback>>,
    on_connected: Mutex<Option<ConnectedCallback>>,
    on_disconnected: Mutex<Option<DisconnectedCallback>>,
}

enum WsCommand {
    Send(Vec<u8>),
    Disconnect,
}

impl WsConnection {
    pub fn new() -> Self {
        Self {
            shared: Arc::new(WsConnectionShared {
                state: AtomicU8::new(TransportState::Disconnected as u8),
                user_initiated_close: AtomicBool::new(false),
                ca_bundle_path: Mutex::new(String::new()),
                on_message: Mutex::new(None),
                on_connected: Mutex::new(None),
                on_disconnected: Mutex::new(None),
            }),
            sender: Mutex::new(None),
            worker: Mutex::new(None),
        }
    }

    pub fn set_ca_bundle_path(&self, path: &str) {
        *self
            .shared
            .ca_bundle_path
            .lock()
            .expect("ws connection poisoned") = path.to_string();
    }

    pub fn ca_bundle_path(&self) -> String {
        self.shared
            .ca_bundle_path
            .lock()
            .expect("ws connection poisoned")
            .clone()
    }

    pub fn state(&self) -> TransportState {
        match self.shared.state.load(Ordering::Relaxed) {
            1 => TransportState::Connecting,
            2 => TransportState::Connected,
            3 => TransportState::Disconnecting,
            _ => TransportState::Disconnected,
        }
    }

    pub fn transition_connecting(&self) -> bool {
        self.shared
            .user_initiated_close
            .store(false, Ordering::Relaxed);
        self.shared
            .state
            .compare_exchange(
                TransportState::Disconnected as u8,
                TransportState::Connecting as u8,
                Ordering::AcqRel,
                Ordering::Acquire,
            )
            .is_ok()
    }

    pub fn mark_connected(&self) {
        self.shared.mark_connected();
    }

    pub fn mark_disconnected(&self, close_code: u16, detail: &str, tls_failed: bool) {
        self.shared
            .mark_disconnected(close_code, detail, tls_failed);
    }

    pub fn mark_user_disconnect(&self) {
        self.shared
            .user_initiated_close
            .store(true, Ordering::Relaxed);
        self.shared
            .state
            .store(TransportState::Disconnecting as u8, Ordering::Release);
    }

    pub fn deliver_binary(&self, data: &[u8]) {
        self.shared.deliver_binary(data);
    }

    pub fn set_on_message<F>(&self, cb: F)
    where
        F: Fn(&[u8]) + Send + Sync + 'static,
    {
        *self
            .shared
            .on_message
            .lock()
            .expect("ws connection poisoned") = Some(Arc::new(cb));
    }

    pub fn set_on_connected<F>(&self, cb: F)
    where
        F: Fn() + Send + Sync + 'static,
    {
        *self
            .shared
            .on_connected
            .lock()
            .expect("ws connection poisoned") = Some(Arc::new(cb));
    }

    pub fn set_on_disconnected<F>(&self, cb: F)
    where
        F: Fn(TransportDisconnectReason, &str) + Send + Sync + 'static,
    {
        *self
            .shared
            .on_disconnected
            .lock()
            .expect("ws connection poisoned") = Some(Arc::new(cb));
    }

    fn connect_worker(&self, url: &str) -> bool {
        if !self.transition_connecting() {
            return false;
        }
        let (tx, rx) = mpsc::channel();
        *self.sender.lock().expect("ws connection poisoned") = Some(tx);
        let shared = Arc::clone(&self.shared);
        let url = url.to_string();
        let worker = thread::spawn(move || run_ws_worker(shared, url, rx));
        *self.worker.lock().expect("ws connection poisoned") = Some(worker);
        true
    }
}

impl Transport for WsConnection {
    fn connect(&mut self, url: &str) -> bool {
        self.connect_worker(url)
    }

    fn send(&mut self, data: &[u8]) -> bool {
        if self.state() != TransportState::Connected {
            return false;
        }
        self.sender
            .lock()
            .expect("ws connection poisoned")
            .as_ref()
            .is_some_and(|tx| tx.send(WsCommand::Send(data.to_vec())).is_ok())
    }

    fn disconnect(&mut self) {
        if self.state() == TransportState::Disconnected {
            return;
        }
        self.mark_user_disconnect();
        let sent = self
            .sender
            .lock()
            .expect("ws connection poisoned")
            .as_ref()
            .is_some_and(|tx| tx.send(WsCommand::Disconnect).is_ok());
        if !sent {
            self.mark_disconnected(NORMAL_CLOSE_CODE, "client disconnect", false);
        }
    }

    fn state(&self) -> TransportState {
        WsConnection::state(self)
    }

    fn set_on_message(&mut self, cb: Box<dyn Fn(&[u8]) + Send + Sync>) {
        *self
            .shared
            .on_message
            .lock()
            .expect("ws connection poisoned") = Some(Arc::from(cb));
    }

    fn set_on_connected(&mut self, cb: Box<dyn Fn() + Send + Sync>) {
        *self
            .shared
            .on_connected
            .lock()
            .expect("ws connection poisoned") = Some(Arc::from(cb));
    }

    fn set_on_disconnected(
        &mut self,
        cb: Box<dyn Fn(TransportDisconnectReason, &str) + Send + Sync>,
    ) {
        *self
            .shared
            .on_disconnected
            .lock()
            .expect("ws connection poisoned") = Some(Arc::from(cb));
    }

    fn set_ca_bundle_path(&mut self, path: &str) {
        WsConnection::set_ca_bundle_path(self, path);
    }
}

impl Default for WsConnection {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for WsConnection {
    fn drop(&mut self) {
        self.disconnect();
    }
}

impl WsConnectionShared {
    fn mark_connected(&self) {
        self.state
            .store(TransportState::Connected as u8, Ordering::Release);
        if let Some(cb) = self
            .on_connected
            .lock()
            .expect("ws connection poisoned")
            .clone()
        {
            cb();
        }
    }

    fn mark_disconnected(&self, close_code: u16, detail: &str, tls_failed: bool) {
        self.state
            .store(TransportState::Disconnected as u8, Ordering::Release);
        let reason = if self.user_initiated_close.load(Ordering::Relaxed) {
            TransportDisconnectReason::UserInitiated
        } else {
            map_close_reason(close_code, tls_failed)
        };
        if let Some(cb) = self
            .on_disconnected
            .lock()
            .expect("ws connection poisoned")
            .clone()
        {
            cb(reason, detail);
        }
    }

    fn deliver_binary(&self, data: &[u8]) {
        if let Some(cb) = self
            .on_message
            .lock()
            .expect("ws connection poisoned")
            .clone()
        {
            cb(data);
        }
    }
}

fn run_ws_worker(shared: Arc<WsConnectionShared>, url: String, rx: mpsc::Receiver<WsCommand>) {
    let (mut socket, _) = match connect(url.as_str()) {
        Ok(connection) => connection,
        Err(err) => {
            let detail = err.to_string();
            let reason = map_error_reason(&detail, false);
            shared.mark_disconnected(reason_to_close_code(reason), &detail, false);
            return;
        }
    };
    let _ = set_read_timeout(socket.get_mut(), Some(READ_POLL_TIMEOUT));
    shared.mark_connected();

    loop {
        while let Ok(cmd) = rx.try_recv() {
            match cmd {
                WsCommand::Send(data) => {
                    if let Err(err) = socket.send(Message::Binary(data.into())) {
                        let detail = err.to_string();
                        let reason = map_error_reason(&detail, false);
                        shared.mark_disconnected(reason_to_close_code(reason), &detail, false);
                        return;
                    }
                }
                WsCommand::Disconnect => {
                    let _ = socket.close(None);
                    shared.mark_disconnected(NORMAL_CLOSE_CODE, "client disconnect", false);
                    return;
                }
            }
        }

        match socket.read() {
            Ok(Message::Binary(bytes)) => shared.deliver_binary(&bytes),
            Ok(Message::Close(frame)) => {
                let detail = frame
                    .as_ref()
                    .map(|frame| frame.reason.to_string())
                    .unwrap_or_else(|| "remote close".to_string());
                let code = frame
                    .map(|frame| u16::from(frame.code))
                    .unwrap_or(NORMAL_CLOSE_CODE);
                shared.mark_disconnected(code, &detail, false);
                return;
            }
            Ok(Message::Ping(_) | Message::Pong(_) | Message::Text(_) | Message::Frame(_)) => {}
            Err(err) if is_timeout_error(&err) => {}
            Err(WsError::ConnectionClosed | WsError::AlreadyClosed) => {
                shared.mark_disconnected(NORMAL_CLOSE_CODE, "remote close", false);
                return;
            }
            Err(err) => {
                let detail = err.to_string();
                let reason = map_error_reason(&detail, false);
                shared.mark_disconnected(reason_to_close_code(reason), &detail, false);
                return;
            }
        }
    }
}

fn set_read_timeout(
    stream: &mut MaybeTlsStream<std::net::TcpStream>,
    timeout: Option<Duration>,
) -> std::io::Result<()> {
    match stream {
        MaybeTlsStream::Plain(stream) => stream.set_read_timeout(timeout),
        MaybeTlsStream::Rustls(stream) => stream.sock.set_read_timeout(timeout),
        #[allow(unreachable_patterns)]
        _ => Ok(()),
    }
}

fn is_timeout_error(err: &WsError) -> bool {
    matches!(err, WsError::Io(io) if matches!(
        io.kind(),
        std::io::ErrorKind::WouldBlock | std::io::ErrorKind::TimedOut
    ))
}

fn reason_to_close_code(reason: TransportDisconnectReason) -> u16 {
    match reason {
        TransportDisconnectReason::RemoteClose | TransportDisconnectReason::UserInitiated => {
            NORMAL_CLOSE_CODE
        }
        _ => 0,
    }
}

pub fn map_close_reason(close_code: u16, tls_handshake_failed: bool) -> TransportDisconnectReason {
    if tls_handshake_failed {
        TransportDisconnectReason::TlsHandshakeFailed
    } else if (1000..1016).contains(&close_code) {
        TransportDisconnectReason::RemoteClose
    } else {
        TransportDisconnectReason::Unknown
    }
}

pub fn map_error_reason(reason: &str, user_initiated: bool) -> TransportDisconnectReason {
    if user_initiated {
        return TransportDisconnectReason::UserInitiated;
    }
    let tls_failed = reason.contains("tls")
        || reason.contains("TLS")
        || reason.contains("SSL")
        || reason.contains("certificate");
    map_close_reason(0, tls_failed)
}

pub fn normal_close_code() -> u16 {
    NORMAL_CLOSE_CODE
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::mpsc;

    #[test]
    fn maps_close_reasons_like_cpp() {
        assert_eq!(
            map_close_reason(0, true),
            TransportDisconnectReason::TlsHandshakeFailed
        );
        assert_eq!(
            map_close_reason(1000, false),
            TransportDisconnectReason::RemoteClose
        );
        assert_eq!(
            map_close_reason(2000, false),
            TransportDisconnectReason::Unknown
        );
        assert_eq!(normal_close_code(), 1000);
        assert_eq!(
            map_error_reason("TLS certificate verify failed", false),
            TransportDisconnectReason::TlsHandshakeFailed
        );
        assert_eq!(
            map_error_reason("connection reset", false),
            TransportDisconnectReason::Unknown
        );
        assert_eq!(
            map_error_reason("connection reset", true),
            TransportDisconnectReason::UserInitiated
        );
    }

    #[test]
    fn transitions_and_callbacks_fire() {
        let ws = WsConnection::new();
        let (tx, rx) = mpsc::channel();
        ws.set_on_connected(move || tx.send("connected").unwrap());
        assert!(ws.transition_connecting());
        assert!(!ws.transition_connecting());
        ws.mark_connected();
        assert_eq!(ws.state(), TransportState::Connected);
        assert_eq!(rx.recv().unwrap(), "connected");
    }
}
